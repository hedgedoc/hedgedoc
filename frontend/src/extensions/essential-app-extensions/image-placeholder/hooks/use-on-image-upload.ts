/*
 * SPDX-FileCopyrightText: 2022 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */
import { useRendererToEditorCommunicator } from '../../../../components/editor-page/render-context/renderer-to-editor-communicator-context-provider'
import { CommunicationMessageType } from '../../../../components/render-page/window-post-message-communicator/rendering-message'
import { Logger } from '../../../../utils/logger'
import { FileContentFormat, readFile } from '../../../../utils/read-file'
import { useCallback } from 'react'

const log = new Logger('useOnImageUpload')

/**
 * Provides a callback that sends a {@link File file} to the editor via iframe communication.
 *
 * @param lineIndex The index of the line in the markdown content where the placeholder is defined
 * @param placeholderIndexInLine The index of the placeholder in the markdown content line
 */
export const useOnImageUpload = (
  lineIndex: number | undefined,
  placeholderIndexInLine: number | undefined
): ((file: File) => void) => {
  const communicator = useRendererToEditorCommunicator()

  return useCallback(
    (file: File) => {
      readFile(file, FileContentFormat.DATA_URL)
        .then((dataUri) => {
          communicator.sendMessageToOtherSide({
            type: CommunicationMessageType.IMAGE_UPLOAD,
            dataUri,
            fileName: file.name,
            lineIndex,
            placeholderIndexInLine
          })
        })
        .catch((error: ProgressEvent) => log.error('Error while uploading image', error))
    },
    [communicator, placeholderIndexInLine, lineIndex]
  )
}
