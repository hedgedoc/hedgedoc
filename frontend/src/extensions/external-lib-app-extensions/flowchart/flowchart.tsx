/*
 * SPDX-FileCopyrightText: 2022 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */
import fontStyles from '../../../../global-styles/variables.module.scss'
import { AsyncLoadingBoundary } from '../../../components/common/async-loading-boundary/async-loading-boundary'
import { ShowIf } from '../../../components/common/show-if/show-if'
import type { CodeProps } from '../../../components/markdown-renderer/replace-components/code-block-component-replacer'
import { useDarkModeState } from '../../../hooks/common/use-dark-mode-state'
import { Logger } from '../../../utils/logger'
import { testId } from '../../../utils/test-id'
import React, { useEffect, useRef, useState } from 'react'
import { Alert } from 'react-bootstrap'
import { Trans, useTranslation } from 'react-i18next'
import { useAsync } from 'react-use'

const log = new Logger('FlowChart')

/**
 * Renders a flowchart.
 *
 * @param code The code to render the flowchart.
 * @see https://flowchart.js.org/
 */
export const FlowChart: React.FC<CodeProps> = ({ code }) => {
  const diagramRef = useRef<HTMLDivElement>(null)
  const [syntaxError, setSyntaxError] = useState(false)
  const darkModeActivated = useDarkModeState()

  useTranslation()

  const {
    value: flowchartLib,
    loading,
    error: libLoadingError
  } = useAsync(async () => import(/* webpackChunkName: "flowchart.js" */ 'flowchart.js'), [])

  useEffect(() => {
    if (libLoadingError) {
      log.error('Error while loading flowchart.js', libLoadingError)
    }
  })

  useEffect(() => {
    if (diagramRef.current === null || flowchartLib === undefined) {
      return
    }
    const currentDiagramRef = diagramRef.current
    try {
      const parserOutput = flowchartLib.parse(code)
      parserOutput.drawSVG(currentDiagramRef, {
        'line-width': 2,
        fill: 'none',
        'font-size': 16,
        'line-color': darkModeActivated ? '#ffffff' : '#000000',
        'element-color': darkModeActivated ? '#ffffff' : '#000000',
        'font-color': darkModeActivated ? '#ffffff' : '#000000',
        'font-family': fontStyles['font-family-base']
      })
      setSyntaxError(false)
    } catch (error) {
      log.error('Error while rendering flowchart', error)
      setSyntaxError(true)
    }

    return () => {
      Array.from(currentDiagramRef.children).forEach((value) => value.remove())
    }
  }, [code, darkModeActivated, flowchartLib])

  return (
    <AsyncLoadingBoundary loading={loading || !flowchartLib} componentName={'flowchart.js'} error={!!libLoadingError}>
      <ShowIf condition={syntaxError}>
        <Alert variant={'danger'}>
          <Trans i18nKey={'renderer.flowchart.invalidSyntax'} />
        </Alert>
      </ShowIf>
      <div ref={diagramRef} {...testId('flowchart')} className={'text-center'} />
    </AsyncLoadingBoundary>
  )
}
