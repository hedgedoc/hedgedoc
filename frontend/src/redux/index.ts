/*
 * SPDX-FileCopyrightText: 2021 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */
import { isDevMode } from '../utils/test-modes'
import { configureStore } from '@reduxjs/toolkit'
import { darkModeReducer } from './dark-mode/slice'
import { editorConfigReducer } from './editor-config/slice'
import { userReducer } from './user/slice'
import { rendererStatusReducer } from './renderer-status/slice'
import { realtimeStatusReducer } from './realtime/slice'
import { historyReducer } from './history/slice'
import { noteDetailsReducer } from './note-details/slice'

export const store = configureStore({
  reducer: {
    darkMode: darkModeReducer,
    editorConfig: editorConfigReducer,
    user: userReducer,
    rendererStatus: rendererStatusReducer,
    realtimeStatus: realtimeStatusReducer,
    history: historyReducer,
    noteDetails: noteDetailsReducer
  },
  devTools: isDevMode
})

export type ApplicationState = ReturnType<typeof store.getState>

export const getGlobalState = (): ApplicationState => store.getState()
