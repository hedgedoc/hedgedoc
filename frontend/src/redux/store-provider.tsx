'use client'
/*
 * SPDX-FileCopyrightText: 2022 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */
import { store } from './index'
import type { PropsWithChildren } from 'react'
import React from 'react'
import { Provider } from 'react-redux'

/**
 * Sets the redux store for the children components.
 *
 * @param children The child components that should access the redux store
 */
export const StoreProvider: React.FC<PropsWithChildren<unknown>> = ({ children }) => {
  return <Provider store={store}>{children}</Provider>
}
