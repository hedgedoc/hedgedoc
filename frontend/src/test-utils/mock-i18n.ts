/*
 * SPDX-FileCopyrightText: 2023 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */
import type { TFunction } from 'i18next'
import i18n from 'i18next'
import { initReactI18next } from 'react-i18next'

/**
 * Initializes i18n with minimal settings and without any data, so it just returns the used key as translation.
 *
 * @return A promise that resolves if i18n has been initialized
 */
export const mockI18n = (): Promise<TFunction> => {
  return i18n.use(initReactI18next).init({
    lng: 'en',
    fallbackLng: 'en',
    ns: ['translationsNS'],
    defaultNS: 'translationsNS',
    interpolation: {
      escapeValue: false
    },
    resources: { en: { translationsNS: {} } }
  })
}
