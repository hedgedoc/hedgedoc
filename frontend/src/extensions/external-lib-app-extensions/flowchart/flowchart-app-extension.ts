/*
 * SPDX-FileCopyrightText: 2023 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */
import type { CheatsheetExtension } from '../../../components/cheatsheet/cheatsheet-extension'
import {
  basicCompletion,
  codeFenceRegex
} from '../../../components/editor-page/editor-pane/autocompletions/basic-completion'
import type { MarkdownRendererExtension } from '../../../components/markdown-renderer/extensions/_base-classes/markdown-renderer-extension'
import { AppExtension } from '../../_base-classes/app-extension'
import { FlowchartMarkdownExtension } from './flowchart-markdown-extension'
import type { CompletionSource } from '@codemirror/autocomplete'

/**
 * Adds support for flow charts to the markdown rendering.
 */
export class FlowchartAppExtension extends AppExtension {
  buildMarkdownRendererExtensions(): MarkdownRendererExtension[] {
    return [new FlowchartMarkdownExtension()]
  }

  buildCheatsheetExtensions(): CheatsheetExtension[] {
    return [{ i18nKey: 'flowchart', categoryI18nKey: 'charts', readMoreUrl: new URL('https://flowchart.js.org/') }]
  }

  buildAutocompletion(): CompletionSource[] {
    return [basicCompletion(codeFenceRegex, '```flow\n\n```')]
  }
}
