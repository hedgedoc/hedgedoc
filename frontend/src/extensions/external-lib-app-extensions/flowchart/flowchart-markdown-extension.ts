/*
 * SPDX-FileCopyrightText: 2022 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */
import { CodeBlockMarkdownRendererExtension } from '../../../components/markdown-renderer/extensions/_base-classes/code-block-markdown-extension/code-block-markdown-renderer-extension'
import { CodeBlockComponentReplacer } from '../../../components/markdown-renderer/replace-components/code-block-component-replacer'
import { FlowChart } from './flowchart'

/**
 * Adds support for flow charts to the markdown rendering using code fences with "flow" as language.
 */
export class FlowchartMarkdownExtension extends CodeBlockMarkdownRendererExtension {
  public buildReplacers(): CodeBlockComponentReplacer[] {
    return [new CodeBlockComponentReplacer(FlowChart, 'flow')]
  }
}
