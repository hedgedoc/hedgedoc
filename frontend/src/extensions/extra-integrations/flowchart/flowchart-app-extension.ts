/*
 * SPDX-FileCopyrightText: 2022 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */
import type { CheatsheetExtension } from '../../../components/editor-page/cheatsheet/cheatsheet-extension'
import type { MarkdownRendererExtension } from '../../../components/markdown-renderer/extensions/base/markdown-renderer-extension'
import { AppExtension } from '../../base/app-extension'
import { FlowchartMarkdownExtension } from './flowchart-markdown-extension'

/**
 * Adds support for flow charts to the markdown rendering.
 */
export class FlowchartAppExtension extends AppExtension {
  buildMarkdownRendererExtensions(): MarkdownRendererExtension[] {
    return [new FlowchartMarkdownExtension()]
  }

  buildCheatsheetExtensions(): CheatsheetExtension[] {
    return [{ i18nKey: 'flowchart', categoryI18nKey: 'charts', readMoreUrl: new URL('https://flowchart.js.org/') }]
  }
}
