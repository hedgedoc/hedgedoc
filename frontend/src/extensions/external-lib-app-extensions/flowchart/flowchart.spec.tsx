/*
 * SPDX-FileCopyrightText: 2023 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */
import { StoreProvider } from '../../../redux/store-provider'
import { mockI18n } from '../../../test-utils/mock-i18n'
import { FlowChart } from './flowchart'
import * as useMediaQuery from '@restart/hooks/useMediaQuery'
import { render, screen } from '@testing-library/react'
import type * as flowchartJsModule from 'flowchart.js'
import type { PropsWithChildren } from 'react'
import React from 'react'

jest.mock('@restart/hooks/useMediaQuery')
jest.mock('../../../components/common/application-error-alert/application-error-alert', () => ({
  ApplicationErrorAlert: ({ children, ...props }: PropsWithChildren) => (
    <div>
      <h3>This is a mock for ApplicationErrorAlert.</h3>
      Props: <code>{JSON.stringify(props)}</code>
      Children:
      <div>{children}</div>
    </div>
  )
}))

describe('Flowchart', () => {
  const successText = 'Flowchart rendering succeeded!'
  const expectedValidFlowchartCode = 'test code'

  beforeAll(async () => {
    jest.spyOn(useMediaQuery, 'default').mockImplementation(() => false)
    await mockI18n()
  })

  afterEach(() => {
    jest.resetModules()
    jest.restoreAllMocks()
  })

  const mockFlowchartDraw = (): jest.Mock<void, Parameters<flowchartJsModule.Instance['drawSVG']>> => {
    const drawSvg = jest.fn((container: HTMLElement | string) => {
      if (typeof container === 'string') {
        throw new Error('HTMLElement expected')
      } else {
        container.innerHTML = successText
      }
    })
    jest.mock('flowchart.js', () => ({
      parse: jest.fn((code) => {
        if (code !== expectedValidFlowchartCode) {
          throw new Error('invalid flowchart code')
        }
        return { drawSVG: drawSvg, clean: jest.fn() }
      })
    }))
    return drawSvg
  }

  it('renders correctly', async () => {
    const successText = 'Flowchart rendering succeeded!'
    const validFlowchartCode = 'test code'
    const mockDrawSvg = mockFlowchartDraw()

    const view = render(
      <StoreProvider>
        <FlowChart code={validFlowchartCode} />
      </StoreProvider>
    )
    await screen.findByText(successText)
    expect(view.container).toMatchSnapshot()
    expect(mockDrawSvg).toBeCalled()
  })

  it('handles error while rendering', async () => {
    const mockDrawSvg = mockFlowchartDraw()

    const view = render(
      <StoreProvider>
        <FlowChart code={'Invalid!'} />
      </StoreProvider>
    )
    await screen.findByText('renderer.flowchart.invalidSyntax')
    expect(view.container).toMatchSnapshot()
    expect(mockDrawSvg).not.toBeCalled()
  })

  it('handles error if lib loading failed', async () => {
    jest.mock('flowchart.js', () => {
      throw new Error('flowchart.js import is exploded!')
    })

    const view = render(
      <StoreProvider>
        <FlowChart code={'Invalid!'} />
      </StoreProvider>
    )
    await screen.findByText('common.errorWhileLoading')
    expect(view.container).toMatchSnapshot()
  })
})
