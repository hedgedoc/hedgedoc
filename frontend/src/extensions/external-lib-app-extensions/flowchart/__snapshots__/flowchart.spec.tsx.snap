// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Flowchart handles error if lib loading failed 1`] = `
<div>
  <div
    class="fade alert alert-danger show"
    role="alert"
  >
    common.errorWhileLoading
  </div>
</div>
`;

exports[`Flowchart handles error while rendering 1`] = `
<div>
  <div
    class="fade alert alert-danger show"
    role="alert"
  >
    renderer.flowchart.invalidSyntax
  </div>
  <div
    class="text-center"
    data-testid="flowchart"
  />
</div>
`;

exports[`Flowchart renders correctly 1`] = `
<div>
  <div
    class="text-center"
    data-testid="flowchart"
  >
    Flowchart rendering succeeded!
  </div>
</div>
`;
